zkevm-circuits/src/evm_circuit/util/common_gadget.rs

```diff
@@ -76,6 +79,160 @@ impl<F: Field> SameContextGadget<F> {
     }
 }
 
+/// Construction of step state transition that restores caller's state.
+#[derive(Clone, Debug)]
+pub(crate) struct RestoreContextGadget<F> {
```

Comment:
I think this gadget can also be used in error cases handling (e.g.  OOG) as well as `STOP`,  because all error cases will trigger call context switch.

Reply:
Yes, tho I haven't taken too much consideration of exceptions on this gadget, but it should cover the basic usage of them.

Reply:
thanks, I will try to implement the OOG cases first base this change .